### DISCUSSION TITLE
How to run an action when a component is removed

### DISCUSSION THREAD

**Discussion by jhasse** [skypjack/entt#1191] [Q&A]
I have the following component:

```cpp
struct BodyComponent {
	b2Body* body;
};
```

Now whenever that component is removed I want to execute the following code:

```cpp
body->GetWorld()->DestroyBody(body);
```

How would I do that given that I can't make `BodyComponent` move-only and do it in the destructor? Am I thinking about ECS completely wrong?

**Comment by skypjack** [comment]
Why you can't make `BodyComponent` move-only? Actually, properly defining move ctor/op here would solve the problem.
This is what I have in a project of mine for example:
```cpp
struct rigid_body {
    rigid_body(b2Body *);

    rigid_body(rigid_body &&) = default;
    rigid_body &operator=(rigid_body &&) = default;

    // ...

private:
    std::unique_ptr<b2Body, void (*)(void *)> body;
};

struct physics_component {
    physics_component(b2Body * /*, other args */);

    physics_component(physics_component &&) = default;
    physics_component &operator=(physics_component &&) = default;

    // ...

    rigid_body body;
    // ...
};
```
Then I have something like this in the cpp file:
```cpp
namespace internal {

static void b2body_deleter(void *ptr) noexcept {
    b2Body *body = static_cast<b2Body *>(ptr);
    b2World *world = body->GetWorld();
    world->DestroyBody(body);
}

} // namespace internal

// ...

rigid_body::rigid_body(b2Body *elem)
    : body{elem, &internal::b2body_deleter} {}

physics_component::physics_component(b2Body *elem /*, other args */)
    : body{elem}
      /* , other members */ {}

// ...
```
Otherwise you can listen on the desctruction of the `BodyComponent` and attach a listener that does something similar:
```cpp
my_registry.on_destroy<BodyComponent>().connect<&my_cleanup_function>();
```
You can find more details in the documentation for this stuff. I prefer the first method but both work at the end of the day.

**Comment by jhasse** [comment]
Oh, I had a lambda where I used `BodyComponent` via pass-by-value which caused a huge compile error so that I assumed having move-only components isn't supported - but you're right, it works perfectly! Thank you very much :)
